pdc/apps/release/models.py

```diff
@@ -118,7 +118,42 @@ def export(self):
         }
 
 
-class ProductVersion(models.Model):
+class AllowedPushTargetsModel(models.Model):
+    """
+    Abstract model providing allowed_push_targets field.
+
+    This field stored in database as mask, masking fields from allowed push
+    targets from parent.
+
+    Abstract method _parent_allowed_push_targets() must be overridden.
+    """
+
+    masked_push_targets = models.ManyToManyField(PushTarget)
+
+    class Meta:
+        abstract = True
+
+    def _parent_allowed_push_targets(self):
+        """
+        Returns allowed push targets from parent.
+
+        This method must be overridden.
+        """
+        raise NotImplementedError()
+
+    @property
+    def allowed_push_targets(self):
+        return self._parent_allowed_push_targets().exclude(id__in=self.masked_push_targets.all())
+
```

Comment:
Currently it lists the parent push_targets when the current masked_push_targets is [].
Like, release's masked_push_targets is [], will list product_version's masked_push_targets

Do we need do this? I think we should just return []

Reply:
If `masked_push_targets` is `[]`, this function basically returns parent's `allowed_push_targets`. Not sure what you mean by returning `[]`.

Reply:
I mean if the masked_push_targets is [], just return the [], not return the parent's allowed_push_targets. I think that's the right way, what's your thinking?

Reply:
Look at the codes and tests again, this just  happens in the first time parent's push_targets change. Make sense to me.

Reply:
Internal data `masked_push_targets` are for masking/hiding allowed push targets in parent. The property name above is `allowed_push_targets` so it showed unmasked targets. Basically it recurses to get product's `allowed_push_targets` and mask it with all `masked_push_targets` between product and current level of recursion.